distributed/metrics.py

```diff
@@ -103,3 +108,168 @@ def resync(self) -> None:
+
+    def __init__(self):
+        self._callbacks = ContextVar(f"MetricHook<{id(self)}>._callbacks", default=[])
+
+    @contextmanager
+    def add_callback(
+        self, callback: Callable[[str, float, str], None]
+    ) -> Iterator[None]:
+        """Add a callback when entering the context and remove it when exiting it.
+        The callback must accept the same parameters as :meth:`digest_metric`.
+        """
+        cbs = self._callbacks.get()
+        tok = self._callbacks.set(cbs + [callback])
+        try:
+            yield
+        finally:
+            tok.var.reset(tok)
+
+    def digest_metric(self, label: str, value: float, unit: str) -> None:
+        """Invoke the currently set context callbacks for an arbitrary quantitative
+        metric.
+        """
+        cbs = self._callbacks.get()
+        for cb in cbs:
+            cb(label, value, unit)
+
+    @contextmanager
+    def meter(
+        self,
+        label: str,
+        unit: str = "seconds",
+        func: Callable[[], float] = timemod.perf_counter,
+        floor: float | Literal[False] = 0.0,
+    ) -> Any:
+        """Convenience context manager which calls func() before and after the wrapped
+        code, calculates the delta, and finally calls :meth:`digest_metric`. It also
+        subtracts any other calls to :meth:`meter` or :meth:`digest_metric` with the
+        same unit performed within the context, so that the total is strictly additive.
+
+        Parameters
+        ----------
+        label: str
+            label to pass to the callback
+        unit: str, optional
+            unit to pass to the callback. Default: seconds
+        func: callable
+            see :func:`meter`
+        floor: bool, optional
+            see :func:`meter`
+        """
+        offsets = []
+
+        def callback(label2: str, value2: float, unit2: str) -> None:
+            if unit2 == unit:
+                # This must be threadsafe to support callbacks invoked from
+                # distributed.utils.offload; '+=' on a float would not be threadsafe!
+                offsets.append(value2)
+
+        try:
+            with self.add_callback(callback), meter(func, floor=False) as m:
+                yield
+        finally:
+            delta = m.delta - sum(offsets)
+            if floor is not False:
+                delta = max(floor, delta)
+            self.digest_metric(label, delta, unit)
+
+    def meter_function(
+        self,
+        label: str,
+        unit: str = "seconds",
+        func: Callable[[], float] = timemod.perf_counter,
+        floor: float | Literal[False] = 0.0,
+    ) -> Callable[[Callable], Callable]:
+        """Same as :meth:`meter`, but as a function decorator"""
+
+        def decorator(f2):
+            @wraps(f2)
+            def wrapper(*args, **kwargs):
+                with self.meter(label, unit, func, floor):
+                    return f2(*args, **kwargs)
+
+            return wrapper
+
+        return decorator
```

Comment:
I tried really hard to have a single method that can be used either as a decorator or as a context manager but not only it looked very messy it was also subtly broken.

Reply:
[`@contextlib.contextmanager`](https://github.com/coiled/platform/discussions/categories) automatically makes something that can both be used as a contextmanager and a decorator:

> [contextmanager()](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) uses [ContextDecorator](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) so the context managers it creates can be used as decorators as well as in [with](https://docs.python.org/3/reference/compound_stmts.html#with) statements. When used as a decorator, a new generator instance is implicitly created on each function call (this allows the otherwise “one-shot” context managers created by [contextmanager()](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) to meet the requirement that context managers support multiple invocations in order to be used as decorators).

Maybe there's a subtle reason why just `@meter` wouldn't work though.

Reply:
O_o I didn't know. Never stop learning...

Reply:
I only recently learned this from one of @graingert's PRs!

Reply:
The trick works with sync functions, but not with async ones :cry: 

```python
from contextlib import contextmanager

@contextmanager
def w():
    print("start")
    yield
    print("end")

@w()
def f():
    print("inside(sync)")

@w()
async def g():
    print("inside(async)")

f()
print("===")
await g()
```
```

start
inside(sync)
end
===
start
end
inside(async)
```

Reply:
Unfortunately, `async` functions require the `asynccontextmanager`:

```
from contextlib import asynccontextmanager, contextmanager
@contextmanager
def w():
    print("start")
    yield
    print("end")

@asynccontextmanager
async def aw():
    print("start")
    yield
    print("end")

    
@w()
def f():
    print("inside(sync)")

@aw()
async def g():
    print("inside(async)")

f()
print("===")
await g()
```

```
start
inside(sync)
end
===
start
inside(async)
end
```